apps/server/src/modules/provisioning/strategy/tsp/tsp.strategy.ts

```diff
@@ -1,24 +1,79 @@
-import { Injectable, NotImplementedException } from '@nestjs/common';
+import { Injectable } from '@nestjs/common';
+import { RoleName } from '@shared/domain/interface';
 import { SystemProvisioningStrategy } from '@shared/domain/interface/system-provisioning.strategy';
-import { OauthDataDto, OauthDataStrategyInputDto, ProvisioningDto } from '../../dto';
+import { IdTokenExtractionFailureLoggableException } from '@src/modules/oauth/loggable';
+import { SchoolService } from '@src/modules/school';
+import { validate } from 'class-validator';
+import jwt, { JwtPayload } from 'jsonwebtoken';
+import {
+	ExternalClassDto,
+	ExternalSchoolDto,
+	ExternalUserDto,
+	OauthDataDto,
+	OauthDataStrategyInputDto,
+	ProvisioningDto,
+	ProvisioningSystemDto,
+} from '../../dto';
 import { TspProvisioningService } from '../../service/tsp-provisioning.service';
 import { ProvisioningStrategy } from '../base.strategy';
 import { BadDataLoggableException } from '../loggable';
+import { TspJwtPayload } from './tsp.jwt.payload';
 
 @Injectable()
 export class TspProvisioningStrategy extends ProvisioningStrategy {
-	constructor(private readonly provisioningService: TspProvisioningService) {
+	constructor(
+		private readonly provisioningService: TspProvisioningService,
+		private readonly schoolService: SchoolService
+	) {
 		super();
 	}
 
 	getType(): SystemProvisioningStrategy {
 		return SystemProvisioningStrategy.TSP;
 	}
 
-	// eslint-disable-next-line @typescript-eslint/no-unused-vars
-	override getData(input: OauthDataStrategyInputDto): Promise<OauthDataDto> {
-		// TODO EW-1004
-		throw new NotImplementedException();
+	override async getData(input: OauthDataStrategyInputDto): Promise<OauthDataDto> {
+		const decodedAccessToken: JwtPayload | null = jwt.decode(input.accessToken, { json: true });
+
+		if (!decodedAccessToken || !decodedAccessToken.sub) {
+			throw new IdTokenExtractionFailureLoggableException('sub');
+		}
+
+		const payload = new TspJwtPayload(decodedAccessToken);
+		const errors = await validate(payload);
+
+		if (errors.length > 0) {
+			throw new IdTokenExtractionFailureLoggableException(errors.map((error) => error.property).join(', '));
+		}
+
+		const provisioningSystemDto = new ProvisioningSystemDto({
+			systemId: payload.sid,
+			provisioningStrategy: SystemProvisioningStrategy.TSP,
+		});
+
+		const externalUserDto = new ExternalUserDto({
+			externalId: payload.sub,
+			firstName: payload.personVorname,
+			lastName: payload.personNachname,
+			roles: Object.values(RoleName).filter((role) => payload.ptscListRolle.split(',').includes(role)),
```

Comment:
Are the Roles from TSP the same as our roles? This probably needs a proper mapper here

Reply:
done

Reply:
done